lib/reversetunnel/seek/seek.go

```diff
@@ -0,0 +1,419 @@
+		p.stopGroup(g)
+	}
+}
+
+// Shutdown stops all seek operations
+func (p *Pool) Shutdown() {
+	p.m.Lock()
+	defer p.m.Unlock()
+	for g, _ := range p.groups {
+		p.stopGroup(g)
+	}
+}
+
+func (p *Pool) stopGroup(key Key) {
+	group, ok := p.groups[key]
+	if !ok {
+		return
+	}
+	group.cancel()
+	delete(p.groups, key)
+}
+
+// Group is a shared handle to a group-level seek manager.
+type Group struct {
+	inner  *proxySeekers
+	cancel context.CancelFunc
+	proxyC chan<- string
+	seekC  <-chan Key
+	statC  <-chan Status
+}
+
+func newGroup(ctx context.Context, conf Config, seekC chan Key, id Key) Group {
+	ctx, cancel := context.WithCancel(ctx)
+	proxyC := make(chan string, 128)
+	statC := make(chan Status, 1)
+	seekers := &proxySeekers{
+		id:     id,
+		conf:   conf,
+		states: make(map[string]seeker),
+		proxyC: proxyC,
+		seekC:  seekC,
+		statC:  statC,
+	}
+	handle := Group{
+		inner:  seekers,
+		cancel: cancel,
+		proxyC: proxyC,
+		seekC:  seekC,
+		statC:  statC,
+	}
+	go seekers.run(ctx)
+	return handle
+}
+
+// WithProxy is used to wrap the connection-handling logic of an agent,
+// ensuring that it is run if and only if no other agent is already
+// handling this proxy.
+func (s *Group) WithProxy(do func(), principals ...string) (did bool) {
+	if !s.inner.TryAcquireProxy(principals...) {
+		return false
+	}
+	defer s.inner.ReleaseProxy(principals...)
+	do()
+	return true
+}
+
+// Status channel is regularly updated with the most recent status
+// value.  Consuming status values is optional.
+func (s *Group) Status() <-chan Status {
+	return s.statC
+}
+
+// Gossip channel must be informed whenever a proxy's identity
+// becomes known via gossip messages.
+func (s *Group) Gossip() chan<- string {
+	return s.proxyC
+}
+
+// proxySeekers manages all proxy seekers for a group.
+type proxySeekers struct {
+	sync.Mutex
+	id       Key
+	conf     Config
+	states   map[string]seeker
+	proxyC   <-chan string
+	seekC    chan<- Key
+	statC    chan Status
+	lastStat *Status
+	lastTick time.Time
+}
+
+func (p *proxySeekers) run(ctx context.Context) {
+	ticker := time.NewTicker(p.conf.TickRate)
+	defer ticker.Stop()
+	// supply initial status & seek notification.
+	p.notifyStatus(p.Tick())
+	p.notifyShouldSeek()
+Run:
+	for {
+		select {
+		case <-ticker.C:
+			stat := p.Tick()
+			p.notifyStatus(stat)
+			if stat.ShouldSeek() {
+				p.notifyShouldSeek()
+			}
+		case proxy := <-p.proxyC:
+			proxies := []string{proxy}
+			// Collect any additional proxy messages
+			// without blocking.
+		Collect:
+			for {
+				select {
+				case pr := <-p.proxyC:
+					proxies = append(proxies, pr)
+				default:
+					break Collect
+				}
+			}
+			count := p.RefreshProxy(proxies...)
+			if count > 0 {
+				p.notifyShouldSeek()
+			}
+		case <-ctx.Done():
+			break Run
+		}
+	}
+}
+
+func (p *proxySeekers) Tick() Status {
+	p.Lock()
+	defer p.Unlock()
+	now := time.Now()
+	return p.tick(now)
+}
+
+func (p *proxySeekers) RefreshProxy(proxies ...string) int {
```

Comment:
Add comment about what this method does and returns (returns a number of proxies that have been refreshed)